fdbclient/NativeAPI.actor.cpp

```diff
@@ -3124,38 +3129,39 @@ ACTOR Future<Void> warmRange_impl(Reference<TransactionState> trState, KeyRange
 	return Void();
 }
 
-SpanID generateSpanID(bool transactionTracingSample, SpanID parentContext = SpanID()) {
-	uint64_t txnId = deterministicRandom()->randomUInt64();
+// TODO - ljoswiak and mpilman. This will need some close inspection.
+SpanContext generateSpanID(bool transactionTracingSample, SpanContext parentContext = SpanContext()) {
+	UID txnId = UID(deterministicRandom()->randomUInt64(), deterministicRandom()->randomUInt64());
 	if (parentContext.isValid()) {
-		if (parentContext.first() > 0) {
-			txnId = parentContext.first();
+		if (parentContext.traceID.first() > 0) {
+			txnId = parentContext.traceID;
 		}
-		uint64_t tokenId = parentContext.second() > 0 ? deterministicRandom()->randomUInt64() : 0;
-		return SpanID(txnId, tokenId);
+		uint64_t spanId = parentContext.spanID > 0 ? deterministicRandom()->randomUInt64() : 0;
+		return SpanContext(txnId, spanId);
```

Comment:
The goal of this first `if` check was to generate a `SpanContext` if the parent `SpanContext` was being sampled. Since the new `SpanContext` class has a flag to determine whether it is being sampled, I think this function has be simplified to generate a new `SpanContext` with the same trace ID as the parent, a unique span ID, and the sample flag set to true if `parentContext` has the sample flag enabled.

Reply:
Yes, there are two cases where a `SpanContext` is generated that indicates the span should be sampled.

The first is if the parent `SpanContext` is being sampled. In that case, we want to make sure all children are sampled as well. That covers the first part of the `if`.

The second case is where we are starting a brand new transaction. There is a knob that can be set to control what percentage of transactions are traced (er rather, have sampling enabled for their traces). You are right, we are basically rolling the dice to determine whether the transaction should be sampled (but the probability is controlled by us through a knob). As FDB gets integrated more with the calling system, we might just be able to set this knob to 0.